### PR TITLE
migrate from argparse to typer

### PR DESCRIPTION
**Pull request by debugmiller**
### Background

Typer has a better developer experience, nicer help messages, and support for shell completions. 

### Changes

* replace argparse with typer
* replace Dynaconf with typer config file parser and environment variables that match the previous ones 
* [BREAKING] multi value flags no longer accept space-separated parameters, instead each argument must be specified with the flag. For example `--valid-table-names table1 table2` must instead be `--valid-table-names table1 --valid-table-names table2`. The flags affected are: `--query-id`, `--analysis-id`, `--filter`, `--ignore-files`, `--valid-table-names`, `--test-names`

### Testing

* updated numerous existing tests, many of which test the full CLI invocation. In addition, these changes result in many more values being explicitly typed, instead of `Any` typed, which makes for much better type checking.

### CODE REVIEW

File: tests/unit/panther_analysis_tool/test_main.py [hunk 1]
```diff
@@ -25,15 +29,93 @@
...
 print("Using fixtures path:", FIXTURES_PATH)
 
+runner = CliRunner()
+
+
+def mock_test_analysis(tc: TestCase, args: list[str]) -> tuple[int, list[str]]:
+    return mock_test_analysis_results(tc, args)[:2]
+
+
+def mock_test_analysis_results(tc: TestCase, args: list[str]) -> tuple[int, list[str], Result]:
```
Comment (debugmiller) [inline tests/unit/panther_analysis_tool/test_main.py:50]: consider combining all these.

File: tests/unit/panther_analysis_tool/test_main.py [hunk 2]
```diff
@@ -75,7 +157,12 @@ def setUp(self):
         # sqlfluff needs to be able to access its package metadata to work
         self.fs.add_package_metadata("sqlfluff")
 
+        main._DISABLE_PANTHER_EXCEPTION_HANDLER = True
+        # skip the http check
+        main._SKIP_HTTP_VERSION_CHECK = True
```
Comment (debugmiller) [inline tests/unit/panther_analysis_tool/test_main.py:162]: disable the config loader here as well

File: panther_analysis_tool/main.py [hunk 3]
```diff
@@ -1572,594 +1715,606 @@ def _print_test_result(
...
+        "is an extension of Data Replay and is subject to the same limitations."
+    ),
+)
+def benchmark_command(
+    _filter: FilterType = None,
+    ignore_files: IgnoreFilesType = None,
+    path: PathType = ".",
+    out: OutType = ".",
+    iterations: Annotated[
+        int,
```
[the remark this answers is not in the document]
Comment (maxrichie5) [reply panther_analysis_tool/main.py:2308]: I think the optional part is still needed because you don't have to supply the number of iterations, even though it is always an int. I would imagine typer would force the arg to be supplied when it doesn't need to be. Maybe I am misunderstanding that though

**Review by maxrichie5 — APPROVED** [review]
Reviewed the latest commit and had no notes for it. Just one question outstanding and then I think this is all good!

File: panther_analysis_tool/main.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (debugmiller) [reply panther_analysis_tool/main.py:2308]: The function signature a few lines below this defaults to 50

**Comment by maxrichie5** [comment]
I had Claude review this PR and this is the output:
------------------------------------------------
Code Review for PR #649: Migrate from argparse to typer

  Overview:
  This pull request aims to replace the argparse library with
  typer for command-line argument parsing in the Panther
  Analysis Tool. The change also includes replacing Dynaconf
  with typer's config file parser and environment variables.
  This is a significant refactoring that affects multiple files
   and introduces breaking changes in the CLI interface.

  Analysis of code quality and style:

  1. Modernization and improved developer experience:
    - The migration to typer is a good choice, as it provides
  better developer experience, nicer help messages, and support
   for shell completions.
    - The code is now more type-safe, with many values being
  explicitly typed instead of using Any.
  2. Code structure and organization:
    - The refactoring is thorough and consistent across
  multiple files.
    - The introduction of dataclasses for argument structures
  (e.g., BenchmarkArgs, BulkDeleteArgs) improves code
  readability and maintainability.
  3. Type annotations:
    - Extensive use of type annotations, including the creation
   of custom types (e.g., APITokenType, FilterType), enhances
  code clarity and helps catch potential errors early.
  4. Configuration management:
    - The switch from Dynaconf to typer's config file parser
  simplifies the configuration management while maintaining
  functionality.
  5. Error handling and logging:
    - The error handling in the call_and_exit decorator is
  well-implemented, providing informative error messages while
  handling different error formats.

  Specific suggestions for improvements:

  1. In panther_analysis_tool/core/parse.py:
    - Consider using an Enum for the Filter class's key
  attribute to restrict possible values and catch potential
  errors at compile-time.
  2. In panther_analysis_tool/main.py:
    - The call_and_exit decorator is quite complex. Consider
  breaking it down into smaller functions for better
  readability and testability.
  3. In panther_analysis_tool/command/benchmark.py:
    - The BenchmarkArgs dataclass has many attributes. Consider
   grouping related attributes into nested dataclasses for
  better organization.

  Potential issues or risks:

  1. Breaking changes:
    - The PR introduces breaking changes in the CLI interface,
  particularly in how multi-value flags are handled. This may
  require updates to existing scripts or documentation that use
   the tool.
  2. Complexity:
    - The refactoring introduces more complex type annotations
  and dataclasses. While this improves type safety, it may
  increase the learning curve for new contributors.
  3. Testing:
    - Ensure that all new type annotations and dataclasses are
  covered by unit tests to prevent regressions in future
  updates.
  4. Performance:
    - The impact on startup time and memory usage should be
  evaluated, especially for large projects with many
  command-line options.

  Test coverage:

  The PR mentions that numerous existing tests have been
  updated. However, it would be beneficial to:
  1. Add new tests specifically for the typer implementation to
   ensure all new functionality is covered.
  2. Include tests for edge cases in argument parsing and
  configuration file handling.

  Security considerations:

  1. The changes don't appear to introduce any new security
  vulnerabilities.
  2. The use of type annotations and stricter parsing may help
  prevent some classes of injection attacks.

  Overall, this PR represents a significant improvement in the
  project's code quality and developer experience. The
  migration to typer and the extensive use of type annotations
  will likely lead to fewer bugs and easier maintenance in the
  future. However, due to the breaking changes, a careful
  rollout plan and updated documentation will be crucial for a
  smooth transition.
